vendors/st/boards/stm32l475_discovery/ports/common_io/iot_spi.c

```diff
@@ -0,0 +1,679 @@
+    .sOpened       = IOT_SPI_CLOSED,
+};
+
+static IotSPIDescriptor_t xSpi3 =
+{
+    .pxSpiContext  = &xSpiContexts[ 2 ],
+    .xConfig       = { 0 },
+    .xSpiCallback  = NULL,
+    .pvUserContext = NULL,
+    .sOpened       = IOT_SPI_CLOSED,
+};
+/*-----------------------------------------------------------*/
+
+static IotSPIHandle_t const pxSpis[] = { &xSpi1, &xSpi2, &xSpi3 };
+
+/*--------------------API Implementation---------------------*/
+
+IotSPIHandle_t iot_spi_open( int32_t lSpiInstance )
+{
+    IotSPIHandle_t xHandle = NULL;
+
+    if( ( lSpiInstance >= 0 ) && ( lSpiInstance <= SPI_INSTANCE_NUM ) )
+    {
+        xHandle = pxSpis[ lSpiInstance ];
+
+        if( xHandle->sOpened == IOT_SPI_CLOSED )
+        {
+            if( HAL_SPI_Init( xHandle->pxSpiContext->pxSpi ) != HAL_OK )
+            {
+                xHandle = NULL;
+            }
+            else
+            {
+                xHandle->sOpened = IOT_SPI_OPENED;
+            }
+        }
+        else
+        {
+            xHandle = NULL;
+        }
+    }
+
+    return xHandle;
+}
+/*-----------------------------------------------------------*/
+
+void iot_spi_set_callback( IotSPIHandle_t const pxSPIPeripheral,
+                           IotSPICallback_t xCallback,
+                           void * pvUserContext )
+{
+    if( ( pxSPIPeripheral != NULL ) )
+    {
+        pxSPIPeripheral->xSpiCallback = xCallback;
+        pxSPIPeripheral->pvUserContext = pvUserContext;
+    }
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_ioctl( IotSPIHandle_t const pxSPIPeripheral,
+                       IotSPIIoctlRequest_t xSPIRequest,
+                       void * const pvBuffer )
+{
+    int32_t lError = IOT_SPI_INVALID_VALUE;
+
+    if( ( pxSPIPeripheral != NULL ) && ( pvBuffer != NULL ) )
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        switch( xSPIRequest )
+        {
+            case eSPISetMasterConfig:
+
+                if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY )
+                {
+                    lError = IOT_SPI_BUS_BUSY;
+                }
+                else
+                {
+                    /* NOTE: HAL API sends dummy data as the same data in the master buffer. So dummy data does not need to be set in config. */
+                    if( ( ( IotSPIMasterConfig_t * ) pvBuffer )->eSetBitOrder == eSPIMSBFirst )
+                    {
+                        LL_SPI_SetTransferBitOrder( pxSpi->Instance, SPI_FIRSTBIT_MSB );
+                    }
+                    else
+                    {
+                        LL_SPI_SetTransferBitOrder( pxSpi->Instance, SPI_FIRSTBIT_LSB );
+                    }
+
+                    switch( ( ( IotSPIMasterConfig_t * ) pvBuffer )->eMode )
+                    {
+                        case eSPIMode0:
+                            LL_SPI_SetClockPolarity( pxSpi->Instance, SPI_POLARITY_LOW );
```

Comment:
Could these blocks each be collapsed into one call instead of two via logical OR?

Reply:
Actually, I feel this is a bug. The second call should be LL_SPI_SetClockPhase instead of LL_SPI_SetClockPolarity.

Reply:
Good catch! This suggests that we don't have test code for this, though. Do we?

Reply:
Yeah, our test cases still need more coverage, especially ioctl. I have a backlog issue for it. As the scope of this problem is bigger, I wouldn't solve it in this PR.